### ISSUE TITLE
Lint/UselessAccessModifier reports wrong modifier

### ISSUE THREAD

**Issue by daviddavis** [bbatsov/rubocop#3483]
This is probably related to https://github.com/bbatsov/rubocop/issues/3422. It looks like UselessAccessModifier is reporting the wrong line number for the following code:

``` ruby
class HostsControllerTest < ActionController::TestCase
  context 'this is a context' do
    test 'that 1 is 1' do
      assert 1, return_one
    end

    private

    def return_one
      1
    end
  end

  test 'that 2 is 2' do
    assert 1, return_two
  end

  private

  def return_two
    2
  end
end
```

Here's the warning:

```
test.rb:18:3: W: Lint/UselessAccessModifier: Useless private access modifier.
  private
  ^^^^^^^
```

Notice it's reporting line 18 instead of 7.

**Comment by maxjacobson** [comment]
```yaml
Lint/UselessAccessModifier:
  ContextCreatingMethods:
    - context
```

Does this resolve your problem? With this config, the cop shouldn't find any offenses

**Comment by daviddavis** [comment]
@maxjacobson so I want the cop to find the first `private` as useless since it is in fact useless. I think the person who wrote the code in this example was trying to ensure that `return_one` could not be used outside of the context block which is already the case (without having to call `private`).

**Comment by maxjacobson** [comment]
Hm interesting. Do you have a heuristic for how rubocop should know that the first one is useless?

**Comment by daviddavis** [comment]
What about if the access modifier is in a code block (as opposed to a class or module)?

**Comment by maxjacobson** [comment]
Hm. It sort of conflicts with common patterns like rails's [concerning](http://api.rubyonrails.org/v4.1/classes/Module/Concerning.html#method-i-concerning) and class_methods, which let you define blocks where `private` can be meaningful. The `ContextCreatingMethods` config exists to support *that* use-case. I could imagine another config which takes a separate list of methods which shouldn't have any access modifiers in them maybe?

**Comment by daviddavis** [comment]
I guess my question would then be: why does this code pass then when I don't have `ContextCreatingMethods` set?

```ruby
class Test
  context "this is a context" do
    private
    def test
       puts 'ok'
    end
  end
end
```

Another thought: is it maybe possible to have the cop prefer access modifiers in the class/module body over ones inside of blocks?

**Comment by daviddavis** [comment]
Also, this cop does correctly give me an error (it just points to the one I wouldn't expect). So it's quite trivial. I'm happy to close this out if you think it'd be too much work for not enough gain.

**Comment by maxjacobson** [comment]
I think the crux of the issue is that rubocop doesn't know anything about the behavior of methods defined by users or gems (like `context` in that last example) and it's not obvious how it should lint for them, because it can't predict what they're doing with the code defined in the blocks passed to them.

The `ContextCreatingMethods` thing is one way to give it a hint. Without a hint, I kind of wonder if this cop should just ignore them? That might be another way of stating your suggestion to "prefer access modifiers in the class/module body over ones inside of blocks".

> I'm happy to close this out if you think it'd be too much work for not enough gain.

If we can identify a useful change that would improve rubocop, I think it's worth keeping this issue open. For what it's worth, I'm just a random contributor (I added the `ContextCreatingMethods` config when I hit a similar issue).

**Comment by jonas054** [comment]
I took a quick look at the RSpec code and found that `context` and similar methods create classes dynamically and evaluates the give blocks in the scope of those classes. This is just too specialized behavior for RuboCop to be aware of. A new cop in [rubocop-rspec](https://github.com/backus/rubocop-rspec) would make sense, I think.

I'm closing the issue. Let us know if you disagree.
